translates: /docs/hooks-intro.md (#44)
reactjs/nl.reactjs.org: content/docs/hooks-intro.md

---
id: hooks-intro
title: Introductie tot Hooks
permalink: docs/hooks-intro.html
next: hooks-overview.html
---

*Hooks* zijn een nieuwe toevoeging in React 16.8. Ze maken het mogelijk om state en andere React features te gebruiken zonder dat je een class hoeft te schrijven.

```js{4,5}
import React, { useState } from 'react';

function Example() {
  // Declareer een nieuwe state variabele, die we "count" zullen noemen
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>Je hebt {count} keer geklikt</p>
      <button onClick={() => setCount(count + 1)}>

>         Klik op me

Author: Van het spaans:
https://es.reactjs.org/docs/hooks-intro.html

En onze oosterburen:
https://de.reactjs.org/docs/hooks-intro.html

Spaans dus wel comments niet display waarden.
Duits wel beide vertaald.

Ik vind vertalen beter, tenzij haar het voorbeeld wordt verwezen in een "Pen"

      </button>
    </div>
  );
}
```

De nieuwe functie `useState` is de eerste "Hook" waarover we zullen leren, maar dit voorbeeld is maar een teaser. Maak je geen zorgen als het nog niet duidelijk is!

**Je kunt beginnen met leren over Hooks [op de volgende pagina](/docs/hooks-overview.html).** Op die pagina, zullen we verder gaan door uit te leggen waarom we Hooks aan React toevoegen en hoe ze kunnen helpen om goede applicaties te maken.

>Opmerking
>
>React 16.8.0 is de eerste release die Hooks ondersteunt. Vergeet niet alle packages te updaten als je upgrade, ook React DOM. React Native zal Hooks ondersteunen in de volgende stable release.

> ## Introductievideo {#video-introduction}

Author: I think so.

https://www.linguee.nl/engels-nederlands/vertaling/introductory+video.html

Op React Conf 2018 hebben Sophie Alpert en Dan Abramov Hooks geïntroduceerd, gevolgd door Ryan Florence die gedemonstreerd heeft hoe je een applicatie refactored om ze te kunnen gebruiken. Bekijk de video hier:

<br>

<iframe width="650" height="366" src="//www.youtube.com/embed/dpw9EHDh2bM" frameborder="0" allowfullscreen></iframe>

## Geen Breaking Changes {#no-breaking-changes}

> Voor we verder gaan, Hooks zijn:

Author: ```suggestion
Merk voordat we verder gaan op: Hooks zijn:
```

**Er zijn geen plannen om classes te verwijderen uit React.** Je kunt meer lezen over de geleidelijke adoptiestrategie voor Hooks [onderaan](#gradual-adoption-strategy) deze pagina.

**Hooks vervangen niet je kennis van React concepten.** In plaats daarvan bieden Hooks een meer directe API voor de React concepten die je al kent: props, state, context, refs, en lifecycle. Zoals we later zullen laten zien, bieden Hooks ook een nieuwe krachtige manier om ze te combineren.

**Als je enkel wilt beginnen met het leren van Hooks, voel je vrij om [direct naar de volgende pagina te springen!](/docs/hooks-overview.html)** Je kunt ook deze pagina blijven lezen om meer te leren over waarom we Hooks toevoegen, en over hoe we ze gaan gebruiken zonder onze applicaties te herschrijven.

## Motivatie {#motivation}

Hooks lossen een breed scala van schijnbaar ongerelateerde problemen op in React die we tegen zijn gekomen gedurende de vijf jaar dat we tienduizenden componenten geschreven en onderhouden hebben. Of je nu React aan het leren bent, dagelijks gebruikt, of zelfs een andere library prefereert met een vergelijkbaar component model, je zou sommige van deze problemen kunnen herkennen.

> ### Het is moeilijk om logica met state te hergebruiken in componenten {#its-hard-to-reuse-stateful-logic-between-components}

Author: https://github.com/reactjs/nl.reactjs.org/issues/5

But this heading didn't have capitals

> React biedt geen manier om herbruikbaar gedrag aan een component te "knopen" (bijvoorbeeld, om het aan een store te koppelen). Als je al een tijd met React werkt, zou je al bekend kunnen zijn met patronen zoals [render props](/docs/render-props.html) en [higher-order components](/docs/higher-order-components.html) die dat proberen op te lossen. Maar om deze patronen te kunnen gebruiken moet je je componenten herstructureren, wat lastig kan zijn, en het maakt je code moeilijk te volgen. Als je naar een typische React applicatie kijkt met React DevTools, zul je waarschijnlijk een "wrapper hell" van componenten vinden, die omringd zijn met providers, consumers, higher-order componenten, render props, en andere abstracties. Hoewel we die [uit DevTools kunnen filteren](https://github.com/facebook/react-devtools/pull/503), wijst dit op een dieper onderliggend probleem: React heeft een beter principe nodig om logica met state te delen.

Author: Deze suggestion heeft geen wijziging.

> Met Hooks kun je logica met state extraheren uit een component, zodat die onafhankelijk kan worden getest en hergebruikt. **Hooks laten je logica met state hergebruiken zonder aanpassingen in je componenten hiërarchie.** Dit maakt het eenvoudig om Hooks te delen met componenten of met de community.

Author: het slaat op "component", ik heb nagezocht dat het "de component" zou moeten zijn waarnaar je dus zou moeten verwijzen met "die". Denk ik. (Ik worstel er ook meet omdat de meeste literatuur over programmeren het steevast over "het component" heeft.

> Met Hooks kun je logica met state extraheren uit een component, zodat die onafhankelijk kan worden getest en hergebruikt. **Hooks laten je logica met state hergebruiken zonder aanpassingen in je componenten hiërarchie.** Dit maakt het eenvoudig om Hooks te delen met componenten of met de community.

Reviewer: If I'm not mistaken, it refers to 'logica met state' and not 'component'. I personally think it feels more natural, but I may be wrong. Maybe we need a third opinion?

> Met Hooks kun je logica met state extraheren uit een component, zodat die onafhankelijk kan worden getest en hergebruikt. **Hooks laten je logica met state hergebruiken zonder aanpassingen in je componenten hiërarchie.** Dit maakt het eenvoudig om Hooks te delen met componenten of met de community.

Author: You are right about it refering to 'logica met state'.

Discussed this one with my wife, she says we are both wrong...

She says it should be "hij" or "deze".

I like "deze" best:
```suggestion
Met Hooks kun je logica met state extraheren uit een component, zodat deze onafhankelijk kan worden getest en hergebruikt. **Hooks laten je logica met state hergebruiken zonder aanpassingen in je componenten hiërarchie.** Dit maakt het eenvoudig om Hooks te delen met componenten of met de community.
```

> Met Hooks kun je logica met state extraheren uit een component, zodat die onafhankelijk kan worden getest en hergebruikt. **Hooks laten je logica met state hergebruiken zonder aanpassingen in je componenten hiërarchie.** Dit maakt het eenvoudig om Hooks te delen met componenten of met de community.

Reviewer: I agree with your wife, 'deze' is better indeed. 😁

We zullen dit verder bespreken in [Bouw Je Eigen Hooks](/docs/hooks-custom.html).

### Complexe componenten worden moeilijk te begrijpen {#complex-components-become-hard-to-understand}

We hebben vaak componenten moeten onderhouden die eenvoudig begonnen, maar uitgroeiden tot een ononderhoudbare brei van logica met state en neveneffecten. Iedere lifecycle methode bevat vaak een mix van ongerelateerde logica. Bijvoorbeeld, componenten kunnen het ophalen van data uitvoeren in `componentDidMount` en `componentDidUpdate`. Echter, dezelfde `componentDidMount` methode kan ook ongerelateerde logica bevatten om event listeners op te zetten, met opruim-code die uitgevoerd wordt in `componentWillUnmount`. Wederzijds gerelateerde code die altijd gelijktijdig wordt aangepast, wordt dus gescheiden, terwijl volledig ongerelateerde code wordt gecombineerd in één enkele methode. Dit maakt dat er makkelijk bugs en inconsistenties kunnen worden geïntroduceerd.

> In veel gevallen is het niet mogelijk om deze componenten op te splitsen in kleinere componenten omdat de logica met state overal verspreid is. Het is ook moeilijk om ze te testen. Dit is één van de redenen waarom veel mensen verkiezen om React te combineren met een aparte state management library. Maar dat introduceert vaak te veel abstractie, je moet steeds tussen vershillende bestanden heen en weer springen, en het maakt hergebruik van componenten moeilijker.

Reviewer: ```suggestion
In veel gevallen is het niet mogelijk om deze componenten op te splitsen in kleinere componenten omdat de logica met state overal verspreid is. Het is ook moeilijk om ze te testen. Dit is één van de redenen waarom veel mensen verkiezen om React te combineren met een aparte state management library. Maar dat introduceert vaak te veel abstractie, je moet steeds tussen verschillende bestanden heen en weer springen, en het maakt hergebruik van componenten moeilijker.
```

> In veel gevallen is het niet mogelijk om deze componenten op te splitsen in kleinere componenten omdat de logica met state overal verspreid is. Het is ook moeilijk om ze te testen. Dit is één van de redenen waarom veel mensen verkiezen om React te combineren met een aparte state management library. Maar dat introduceert vaak te veel abstractie, je moet steeds tussen verschillende bestanden heen en weer springen, en het maakt hergebruik van componenten moeilijker.

Reviewer 2: ```suggestion
In veel gevallen is het niet mogelijk om deze componenten op te splitsen in kleinere componenten omdat de logica met state overal verspreid is. Het is ook moeilijk om ze te testen. Dit is één van de redenen waarom veel mensen ervoor kiezen om React te combineren met een aparte state management library, maar dat introduceert vaak te veel abstractie. Je moet constant tussen verschillende bestanden heen en weer springen en het maakt hergebruik van componenten moeilijker.
```

> Om dit op te lossen **laten Hooks je een component opsplitsen in kleinere functies gebaseerd op welke onderdelen gerelateerd zijn (zoals het opzetten van een subscription of het ophalen van data)**, in plaats van splitsing te forceren op basis van lifecycle methoden. Je kunt ook opteren om de lokale state van een component te managen met een reducer, om hem voorspelbaarder te maken.

Reviewer 2: ```suggestion
Om dit op te lossen **laten Hooks je een component opsplitsen in kleinere functies gebaseerd op welke onderdelen gerelateerd zijn (zoals het opzetten van een subscription of het ophalen van data)**, in plaats van splitsingen te forceren op basis van lifecycle methoden. Je kunt ook ervoor kiezen om de lokale state van een component te managen met een reducer om hem voorspelbaarder te maken.
```

Author: ```suggestion
Om dit op te lossen **laten Hooks je een component opsplitsen in kleinere functies gebaseerd op welke onderdelen gerelateerd zijn (zoals het opzetten van een subscription of het ophalen van data)**, in plaats van splitsingen te forceren op basis van lifecycle methoden. Je kunt er ook voor kiezen om de lokale state van een component te managen met een reducer, om hem voorspelbaarder te maken.
```

> We zullen dit verder bespreken in [De the Effect Hook](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns).

Reviewer: ```suggestion
We zullen dit verder bespreken in [De Effect Hook](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns).
```

### Classes verwarren zowel mensen als machines {#classes-confuse-both-people-and-machines}

> Behalve dat het hergebruik en code organisatie moeilijke maakt, hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/), is de code erg breedsprakig. Mensen kunnen props, state, en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer welke te gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.

Reviewer: ```suggestion
Naast dat het hergebruik en organisatie van code moeilijker maakt, hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/), is de code erg breedsprakig. Mensen kunnen props, state, en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer welke te gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.
```
Small suggestions for improvement + typo.

> Naast het moeilijker maken van hergebruik en organisatie van code, hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen (Engels)](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/), is de code erg breedsprakig. Mensen kunnen props, state, en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer welke te gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.

Reviewer: ```suggestion
Naast het moeilijker maken van hergebruik en organisatie van code, hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen (Engels)](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/), is de code erg breedsprakig. Mensen kunnen props, state en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer welke te gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.
```
Unnecessary comma.

> Naast het moeilijker maken van hergebruik en organisatie van code, hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen (Engels)](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/), is de code erg breedsprakig. Mensen kunnen props, state en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer welke te gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.

Reviewer 2: ```suggestion
Naast het moeilijker maken van hergebruik en organisatie van code hebben we gemerkt dat classes ook een groot obstakel kunnen zijn bij het leren van React. Je moet begrijpen hoe `this` werkt in JavaScript, wat heel afwijkend is van hoe dat werkt in de meeste andere talen. Je moet er aan denken om de event handlers te binden. Zonder instabiele [syntax voorstellen (Engels)](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/) is de code erg breedsprakig. Mensen kunnen props, state en top-down data flow prima begrijpen maar toch worstelen met classes. Het verschil tussen functie- en class componenten in React en wanneer je welke moet gebruiken leidt tot onenigheid, zelfs onder ervaren React developers.
```

Daarnaast, React is er nu zo'n vijf jaar, en we willen ervoor zorgen dat het de volgende vijf jaar relevant blijft. Zoals [Svelte](https://svelte.technology/), [Angular](https://angular.io/), [Glimmer](https://glimmerjs.com/), en anderen laten zien, heeft [ahead-of-time compilation](https://en.wikipedia.org/wiki/Ahead-of-time_compilation) van componenten veel toekomstig potentieel. Vooral als dat niet gelimiteerd is tot templates. Recentelijk hebben we geëxperimenteerd met [component folding](https://github.com/facebook/react/issues/7323) gebruik makend van [Prepack](https://prepack.io/), en we hebben veelbelovende eerste resultaten gezien. Maar, we vonden dat class componenten onbedoelde patronen kan aanmoedigen, waardoor deze optimalisaties terugvallen naar een langzamer pad. Classes geven ook problemen met tools van vandaag de dag. Bijvoorbeeld, classes worden niet erg goed ge-minified, en ze maken "hot reloading" slordig en onbetrouwbaar. We willen een API presenteren die het aannemelijker maakt dat code op het optimaliseerbare pad blijft.

> Om deze problemen op te lossen, **laten Hooks je meer van React's voorzieningen gebruiken zonder classes.** Conceptueel gezien zijn React componenten altijd al meer een soort functies geweest. Hooks omarmen functies, maar zonder de praktische aard van React op te offeren. Hooks bieden toegang to imperatieve ontsnappingsluiken en vereisen niet dat je complexe functional of reactive programmeertechnieken leert.

Author: Het zou dan denk ik "aangeleund tegen" zijn.

> Om deze problemen op te lossen, **laten Hooks je meer van React's voorzieningen gebruiken zonder classes.** Conceptueel gezien zijn React componenten altijd al meer een soort functies geweest. Hooks omarmen functies, maar zonder de praktische aard van React op te offeren. Hooks bieden toegang to imperatieve ontsnappingsluiken en vereisen niet dat je complexe functional of reactive programmeertechnieken leert.

Author: Ik stel dit dus voor

```suggestion
Om deze problemen op te lossen, **laten Hooks je meer van React's features gebruiken zonder classes.** Conceptueel gezien hebben React componenten altijd al dichter aangeleund tegen  functies. Hooks omarmen functies, maar zonder de praktische aard van React op te offeren. Hooks bieden toegang tot imperatieve ontsnappingsluiken en vereisen niet dat je complexe functional of reactive programmeertechnieken leert.
```

> Om deze problemen op te lossen, **laten Hooks je meer van React's voorzieningen gebruiken zonder classes.** Conceptueel gezien zijn React componenten altijd al meer een soort functies geweest. Hooks omarmen functies, maar zonder de praktische aard van React op te offeren. Hooks bieden toegang to imperatieve ontsnappingsluiken en vereisen niet dat je complexe functional of reactive programmeertechnieken leert.

Reviewer: Good point. I agree with your suggestion above.

>Voorbeelden
>
>[Hooks een Overzicht](/docs/hooks-overview.html) is een goede plek om te beginnen met het leren over Hooks.

## Geleidelijke Adoptie Strategie {#gradual-adoption-strategy}

>**TLDR: Er zijn geen plannen om classes uit React te verwijderen.**

We weten dat React developers gefocused zijn op producten maken en niet de tijd hebben om zich te verdiepen in iedere nieuwe API die wordt uitgebracht. Hooks zijn heel nieuw, en het is misschien beter om te wachten op meer voorbeelden en tutorials, voordat je overweegt ze te leren of te gebruiken.

> We begrijpen ook dat de lat voor het toevoegen van een nieuw principe aan React extreem hoog ligt. Voor nieuwsgierige lezers, hebben we een [uitgebreide RFC](https://github.com/reactjs/rfcs/pull/68) gemaakt die in meer detail duikt in de motivatie, en extra perspectief biedt op de specifieke ontwerpbeslissingen en stand van de techniek.

Reviewer: ```suggestion
We begrijpen ook dat de lat voor het toevoegen van een nieuw principe aan React extreem hoog ligt. Voor nieuwsgierige lezers hebben we een [uitgebreide RFC](https://github.com/reactjs/rfcs/pull/68) gemaakt die in meer detail duikt in de motivatie, en extra perspectief biedt op de specifieke ontwerpbeslissingen en stand van de techniek.
```
